protocols/mdns/src/behaviour.rs

```diff
@@ -304,93 +312,113 @@ where
 
-                        handle.abort();
+                            handle.abort();
+                        }
                     }
+                    Err(err) => tracing::error!("if watch returned an error: {}", err),
                 }
-                Err(err) => tracing::error!("if watch returned an error: {}", err),
             }
-        }
-        // Emit discovered event.
-        let mut discovered = Vec::new();
-
-        while let Poll::Ready(Some((peer, addr, expiration))) =
-            self.query_response_receiver.poll_next_unpin(cx)
-        {
-            if let Some((_, _, cur_expires)) = self
-                .discovered_nodes
-                .iter_mut()
-                .find(|(p, a, _)| *p == peer && *a == addr)
+            // Emit discovered event.
+            let mut discovered = Vec::new();
+
+            while let Poll::Ready(Some((peer, addr, expiration))) =
+                self.query_response_receiver.poll_next_unpin(cx)
             {
-                *cur_expires = cmp::max(*cur_expires, expiration);
-            } else {
-                tracing::info!(%peer, address=%addr, "discovered peer on address");
-                self.discovered_nodes.push((peer, addr.clone(), expiration));
-                discovered.push((peer, addr));
+                if let Some((_, _, cur_expires)) = self
+                    .discovered_nodes
+                    .iter_mut()
+                    .find(|(p, a, _)| *p == peer && *a == addr)
+                {
+                    *cur_expires = cmp::max(*cur_expires, expiration);
+                } else {
+                    tracing::info!(%peer, address=%addr, "discovered peer on address");
+                    self.discovered_nodes.push((peer, addr.clone(), expiration));
+                    discovered.push((peer, addr.clone()));
+
+                    self.pending_events
+                        .push_back(ToSwarm::NewExternalAddrOfPeer {
+                            peer_id: peer,
+                            address: addr,
+                        });
+                }
             }
-        }
 
-        if !discovered.is_empty() {
-            let event = Event::Discovered(discovered);
-            return Poll::Ready(ToSwarm::GenerateEvent(event));
-        }
-        // Emit expired event.
-        let now = Instant::now();
-        let mut closest_expiration = None;
-        let mut expired = Vec::new();
-        self.discovered_nodes.retain(|(peer, addr, expiration)| {
-            if *expiration <= now {
-                tracing::info!(%peer, address=%addr, "expired peer on address");
-                expired.push((*peer, addr.clone()));
-                return false;
+            if !discovered.is_empty() {
+                let event = Event::Discovered(discovered);
+                /// Push to the front of the queue so that the behavior event is reported before
+                /// the individual discovered addresses.
```

Comment:
```suggestion
                // Push to the front of the queue so that the behavior event is reported before
                // the individual discovered addresses.
```